main.go

```diff
@@ -75,57 +85,17 @@ func ProcessExportsSince(wh warehouse.Warehouse, since time.Time) (int, error) {
 		defer outfile.Close()
 		csvOut := csv.NewWriter(outfile)
 
-		stream, err := fs.ExportData(e.ID)
+		recordCount, err := WriteBundleToCSV(fs, e.ID, csvOut, wh)
 		if err != nil {
-			log.Printf("Failed to fetch bundle %d: %s", e.ID, err)
-			return 0, err
-		}
-		defer stream.Close()
-
-		gzstream, err := gzip.NewReader(stream)
-		if err != nil {
-			log.Printf("Failed gzip reader: %s", err)
-			return 0, err
-		}
-
-		decoder := json.NewDecoder(gzstream)
-		decoder.UseNumber()
-
-		// skip array open delimiter
-		if _, err := decoder.Token(); err != nil {
-			log.Printf("Failed json decode of array token: %s", err)
-			return 0, err
-		}
-
-		var recordCount int
-		for decoder.More() {
-			var r Record
-			if err := decoder.Decode(&r); err != nil {
-				log.Printf("failed json decode of record: %s", err)
-				return 0, err
-			}
-			line, err := TransformExportJsonRecord(wh, r)
-			if err != nil {
-				log.Printf("Failed object transform, bundle %d; skipping record. %s", e.ID, err)
-				continue
-			}
-			csvOut.Write(line)
-			recordCount++
-		}
-
-		if _, err := decoder.Token(); err != nil {
-			log.Printf("Failed json decode of array token: %s", err)
 			return 0, err
 		}
 
-		csvOut.Flush()
-
 		if err := wh.LoadToWarehouse(filename); err != nil {
 			log.Printf("Failed to load file '%s' to warehouse: %s", filename, err)
 			return 0, err
 		}
 
-		if err := wh.SaveSyncPoint(e.ID, e.Stop); err != nil {
+		if err := wh.SaveSyncPoints(exports[i : i+1]); err != nil {
```

Comment:
Eww. ;) 

That's tricky to read. I suggest changing `SaveSyncPoint` to a variadic function so it reads cleanly (and you don't even have to change the logic). So like:

`func (rs *Redshift) SaveSyncPoint(bundles ...fullstory.ExportMeta) error `

and then the call here is just: `if err := wh.SaveSyncPoint(e); err != nil {`

and the corresponding call in the grouped version is `wh.SaveSyncPoint(processedBundles...)`

Reply:
Thanks, that's wonderfully elegant.